src/material/schematics/ng-update/migrations/legacy-imports-error.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Rule, SchematicContext, Tree} from '@angular-devkit/schematics';
+import {NodePackageInstallTask} from '@angular-devkit/schematics/tasks';
+import * as ts from 'typescript';
+
+/**
+ * "Migration" that logs an error and prevents further migrations
+ * from running if the project is using legacy components.
+ * @param onSuccess Rule to run if there are no legacy imports.
+ */
+export function legacyImportsError(onSuccess: Rule): Rule {
+  return async (tree: Tree, context: SchematicContext) => {
+    const filesUsingLegacyImports = new Set<string>();
+
+    tree.visit(path => {
+      if (!path.endsWith('.ts')) {
+        return;
+      }
+
+      const content = tree.readText(path);
+      const sourceFile = ts.createSourceFile(path, content, ts.ScriptTarget.Latest);
+
+      sourceFile.forEachChild(function walk(node) {
+        const isImportOrExport = ts.isImportDeclaration(node) || ts.isExportDeclaration(node);
+
+        if (
+          isImportOrExport &&
+          node.moduleSpecifier &&
+          ts.isStringLiteralLike(node.moduleSpecifier) &&
+          node.moduleSpecifier.text.startsWith('@angular/material/legacy-')
+        ) {
+          filesUsingLegacyImports.add(path);
+        }
+
+        node.forEachChild(walk);
+      });
+    });
+
+    // If there are no legacy imports left, we can continue with the migrations.
+    if (filesUsingLegacyImports.size === 0) {
+      return onSuccess;
+    }
+
+    // At this point the project is already at v17 so we need to downgrade it back
+    // to v16 and run `npm install` again. Ideally we would also throw an error here
+    // to interrupt the update process, but that would interrupt `npm install` as well.
+    if (tree.exists('package.json')) {
+      let packageJson: Record<string, any> | null = null;
+
+      try {
+        packageJson = JSON.parse(tree.readText('package.json')) as Record<string, any>;
+      } catch {}
+
+      if (packageJson !== null && packageJson['dependencies']) {
+        packageJson['dependencies']['@angular/material'] = '^16.2.0';
+        tree.overwrite('package.json', JSON.stringify(packageJson, null, 2));
```

Comment:
Note that we have a utility for adding dependencies to the `package.json` (in at least two places!), however my understanding is that we can't put it into a common spot, because the other places are `ng add` schematics and they might not be able to import other packages yet (I may be wrong). This operation is pretty basic so we should be fine.